Source/HTTP/Curl/CurlMulti.cpp

```diff
@@ -0,0 +1,157 @@
+#include "pch.h"
+#include "CurlMulti.h"
+#include "CurlProvider.h"
+
+namespace xbox
+{
+namespace http_client
+{
+
+// XCurl doesn't support curl_multi_timeout, so use a small, fixed delay between calls to curl_multi_perform
+#define PERFORM_DELAY_MS 50
+
+Result<HC_UNIQUE_PTR<CurlMulti>> CurlMulti::Initialize(XTaskQueuePortHandle workPort)
+{
+    assert(workPort);
+
+    http_stl_allocator<CurlMulti> a{};
+    HC_UNIQUE_PTR<CurlMulti> multi{ new (a.allocate(1)) CurlMulti };
+
+    multi->m_curlMultiHandle = curl_multi_init();
+    if (!multi->m_curlMultiHandle)
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "XCurlMulti::Initialize: curl_multi_init failed");
+        return E_FAIL;
+    }
+
+    RETURN_IF_FAILED(XTaskQueueCreateComposite(workPort, workPort, &multi->m_queue));
+
+    return Result<HC_UNIQUE_PTR<CurlMulti>>{ std::move(multi) };
+}
+
+CurlMulti::~CurlMulti()
+{
+    if (m_queue)
+    {
+        XTaskQueueTerminate(m_queue, true, nullptr, nullptr);
+        XTaskQueueCloseHandle(m_queue);
+    }
+
+    if (!m_easyRequests.empty())
+    {
+        HC_TRACE_WARNING(HTTPCLIENT, "XCurlMulti::~XCurlMulti: Failing all active requests.");
+        FailAllRequests(E_UNEXPECTED);
+    }
+
+    if (m_curlMultiHandle)
+    {
+        curl_multi_cleanup(m_curlMultiHandle);
+    }
+}
+
+HRESULT CurlMulti::AddRequest(HC_UNIQUE_PTR<CurlEasyRequest>&& easyRequest)
+{
+    std::unique_lock<std::mutex> lock{ m_mutex };
+
+    auto result = curl_multi_add_handle(m_curlMultiHandle, easyRequest->Handle());
+    if (result != CURLM_OK)
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "XCurlMulti::AddRequest: curl_multi_add_handle failed with CURLCode=%u", result);
+        return HrFromCurlm(result);
+    }
+
+    m_easyRequests.emplace(easyRequest->Handle(), std::move(easyRequest));
+
+    RETURN_IF_FAILED(XTaskQueueSubmitCallback(m_queue, XTaskQueuePort::Work, this, CurlMulti::TaskQueueCallback));
```

Comment:
should release m_mutex before submitting callback in case m_queue is immediate dispatch

Reply:
immediate dispatch won't work at all here and will fail when you try to add a delayed callback?  Should that be checked up front?

Reply:
why did we ever make immediate mode, it adds so many headaches
we should probably just reject immediate queues when the request is being passed in

Reply:
Ya fair point.  Should probably fail for all providers but will add a check here for now

Reply:
Actually seems like we can't even find the dispatch mode of a queue after creation.  That HR should get bubbled up correctly so will just leave it.